# MP004: require-lock-timeout

**Severity**: Critical | **Auto-fixable**: Yes | **Tier**: Free

## What It Detects
DDL operations should set lock_timeout to prevent blocking the lock queue indefinitely.

## Why It's Dangerous
Without lock_timeout, if the table is locked by another query, your DDL waits indefinitely. All subsequent queries pile up behind it in the lock queue, causing cascading timeouts across your application. GoCardless enforces a 750ms lock_timeout for this reason.

## Bad Example
```sql
ALTER TABLE users ADD COLUMN age integer;
```

## Good Example
```sql
SET lock_timeout = '5s';
ALTER TABLE users ADD COLUMN age integer;
RESET lock_timeout;
```

## Auto-fix
`migrationpilot analyze file.sql --fix` prepends `SET lock_timeout = '5s';` before DDL statements automatically.

## Configuration
```yaml
rules:
  MP004: false           # disable
  MP004:
    severity: warning    # change severity
```

## Related Rules
- [MP020](./MP020.md) require-statement-timeout -- statement_timeout for long-running DDL
- [MP019](./MP019.md) no-exclusive-lock-high-connections -- lock contention with many connections
- [MP013](./MP013.md) high-traffic-table-ddl -- DDL on high-traffic tables
