# MP006: no-vacuum-full

**Severity**: Critical | **Auto-fixable**: No | **Tier**: Free

## What It Detects
VACUUM FULL rewrites the entire table under ACCESS EXCLUSIVE lock, blocking all reads and writes.

## Why It's Dangerous
VACUUM FULL rewrites the entire table to a new file, holding an ACCESS EXCLUSIVE lock for the full duration. On large tables this can take hours. Use regular VACUUM or pg_repack for online table compaction instead.

## Bad Example
```sql
VACUUM FULL users;
```

## Good Example
```sql
-- Use pg_repack instead (no ACCESS EXCLUSIVE lock during rewrite):
-- Install: CREATE EXTENSION pg_repack;
-- Run: pg_repack --table users --no-superuser-check

-- Or use regular VACUUM (no exclusive lock):
VACUUM users;
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP006: false           # disable
  MP006:
    severity: warning    # change severity
```

## Related Rules
- [MP032](./MP032.md) ban-cluster -- CLUSTER also rewrites the entire table
- [MP020](./MP020.md) require-statement-timeout -- long-running operations need timeouts
- [MP004](./MP004.md) require-lock-timeout -- set lock_timeout before DDL
