# MP012: no-enum-add-value-in-transaction

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
ALTER TYPE ... ADD VALUE cannot run inside a transaction block on PG < 12. Even on PG 12+, enum modifications take ACCESS EXCLUSIVE on the type.

## Why It's Dangerous
On PostgreSQL versions before 12, ALTER TYPE ADD VALUE inside a transaction raises a runtime error, failing your migration entirely. On PG 12+ it works but takes an ACCESS EXCLUSIVE lock on the type, which can block concurrent queries.

## Bad Example
```sql
BEGIN;
ALTER TYPE status_type ADD VALUE 'archived';
COMMIT;
```

## Good Example
```sql
-- Run outside a transaction block:
ALTER TYPE status_type ADD VALUE 'archived';
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP012: false           # disable
  MP012:
    severity: critical   # change severity
```

## Related Rules
- [MP024](./MP024.md) no-enum-value-removal -- DROP TYPE destroys enum and dependent columns
- [MP025](./MP025.md) ban-concurrent-in-transaction -- other operations that fail inside transactions
- [MP008](./MP008.md) no-multi-ddl-transaction -- avoid wrapping DDL in transactions
