# MP010: no-rename-column

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
RENAME COLUMN breaks application queries referencing the old column name. Prefer adding a new column and migrating data.

## Why It's Dangerous
Renaming a column takes an ACCESS EXCLUSIVE lock and instantly breaks every application query, view, and function referencing the old name. Use the expand-contract pattern: add a new column, migrate data, update code, then drop the old column.

## Bad Example
```sql
ALTER TABLE users RENAME COLUMN name TO full_name;
```

## Good Example
```sql
-- Step 1: Add new column
ALTER TABLE users ADD COLUMN full_name text;

-- Step 2: Backfill data (in batches for large tables)
UPDATE users SET full_name = name WHERE full_name IS NULL;

-- Step 3: Update application code to use "full_name"
-- Step 4: Stop writing to "name"
-- Step 5: Drop old column (after verifying no reads)
ALTER TABLE users DROP COLUMN name;
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP010: false           # disable
  MP010:
    severity: critical   # change severity
```

## Related Rules
- [MP028](./MP028.md) no-rename-table -- renaming a table has similar risks
- [MP007](./MP007.md) no-column-type-change -- type changes also use expand-contract
- [MP017](./MP017.md) no-drop-column -- dropping columns requires careful coordination
