# MP014: large-table-ddl

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Pro

## What It Detects
DDL with long-held locks on a table with over 1M rows. Lock duration will scale with table size.

## Why It's Dangerous
DDL operations on large tables take proportionally longer. Lock duration scales with row count and table size -- what takes seconds on a small table can take minutes or hours on a table with millions of rows.

## Bad Example
```sql
-- Running on a table with 10M rows, 5 GB total size
ALTER TABLE events ALTER COLUMN payload TYPE jsonb;
```

## Good Example
```sql
-- Set a lock_timeout to fail fast:
SET lock_timeout = '5s';
ALTER TABLE events ADD COLUMN payload_new jsonb;
RESET lock_timeout;

-- Backfill in batches, then swap columns
-- Run during maintenance windows for very large tables
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP014: false           # disable
  MP014:
    severity: critical   # change severity
```

## Related Rules
- [MP013](./MP013.md) high-traffic-table-ddl -- DDL on high-traffic tables
- [MP004](./MP004.md) require-lock-timeout -- set lock_timeout before DDL
- [MP020](./MP020.md) require-statement-timeout -- set statement_timeout for long operations
