# MP013: high-traffic-table-ddl

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Pro

## What It Detects
DDL on a table with high query traffic. Lock acquisition may be slow and cause cascading timeouts.

## Why It's Dangerous
Running DDL on tables with high query volume amplifies the blast radius. Even brief locks cause significant query queuing when thousands of queries per second hit the table, leading to cascading timeouts across dependent services.

## Bad Example
```sql
-- Running DDL on a table with 50,000+ queries/minute
ALTER TABLE orders ADD COLUMN priority integer;
```

## Good Example
```sql
-- Set a short lock_timeout to fail fast instead of blocking queries:
SET lock_timeout = '3s';
ALTER TABLE orders ADD COLUMN priority integer;
RESET lock_timeout;

-- Consider running during low-traffic hours.
-- If lock acquisition fails, retry with exponential backoff.
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP013: false           # disable
  MP013:
    severity: critical   # change severity
```

## Related Rules
- [MP014](./MP014.md) large-table-ddl -- DDL on large tables has similar risks
- [MP019](./MP019.md) no-exclusive-lock-high-connections -- lock contention with many connections
- [MP004](./MP004.md) require-lock-timeout -- always set lock_timeout before DDL
