# MP017: no-drop-column

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
DROP COLUMN takes ACCESS EXCLUSIVE lock and may break running application code.

## Why It's Dangerous
DROP COLUMN takes an ACCESS EXCLUSIVE lock and instantly breaks any application code still selecting or inserting that column. Always remove all code references in a prior deploy before dropping the column.

## Bad Example
```sql
ALTER TABLE users DROP COLUMN legacy_field;
```

## Good Example
```sql
-- Safe multi-deploy approach:
-- Deploy 1: Remove all application code references to "legacy_field"
-- Deploy 2: Drop the column with a short lock timeout
SET lock_timeout = '5s';
ALTER TABLE users DROP COLUMN legacy_field;
RESET lock_timeout;
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP017: false           # disable
  MP017:
    severity: critical   # change severity
```

## Related Rules
- [MP010](./MP010.md) no-rename-column -- renaming columns has similar application breakage risks
- [MP026](./MP026.md) ban-drop-table -- dropping tables is even more destructive
- [MP004](./MP004.md) require-lock-timeout -- set lock_timeout before DDL
