# MP029: ban-drop-not-null

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
Dropping NOT NULL allows NULL values in a previously non-nullable column. This may break application code that assumes the column is always populated.

## Why It's Dangerous
Dropping a NOT NULL constraint allows NULL values to be inserted into a column that was previously guaranteed non-null. Application code, ORMs, and downstream systems may crash or return incorrect results when encountering unexpected NULLs.

## Bad Example
```sql
ALTER TABLE users ALTER COLUMN email DROP NOT NULL;
```

## Good Example
```sql
-- Verify that all application code handles NULL values for "email"
-- before dropping NOT NULL. Consider adding a default value instead:
ALTER TABLE users ALTER COLUMN email SET DEFAULT '';
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP029: false           # disable
  MP029:
    severity: critical   # change severity
```

## Related Rules
- [MP002](./MP002.md) require-check-not-null-pattern -- safe pattern for adding NOT NULL
- [MP018](./MP018.md) no-force-set-not-null -- SET NOT NULL needs CHECK constraint pattern
- [MP017](./MP017.md) no-drop-column -- column changes can break application code
