# MP020: require-statement-timeout

**Severity**: Warning | **Auto-fixable**: Yes | **Tier**: Free

## What It Detects
Long-running DDL should have a statement_timeout to prevent holding locks indefinitely.

## Why It's Dangerous
Without statement_timeout, a DDL operation that encounters unexpected conditions (bloated table, heavy WAL, slow I/O) can hold locks for hours, turning a routine migration into a full outage.

## Bad Example
```sql
CREATE INDEX idx_events_created ON events (created_at);
```

## Good Example
```sql
SET statement_timeout = '30s';
CREATE INDEX idx_events_created ON events (created_at);
RESET statement_timeout;
```

## Auto-fix
`migrationpilot analyze file.sql --fix` prepends `SET statement_timeout = '30s';` before long-running DDL statements automatically.

## Configuration
```yaml
rules:
  MP020: false           # disable
  MP020:
    severity: critical   # change severity
```

## Related Rules
- [MP004](./MP004.md) require-lock-timeout -- lock_timeout prevents queue buildup
- [MP006](./MP006.md) no-vacuum-full -- VACUUM FULL is a long-running operation
- [MP032](./MP032.md) ban-cluster -- CLUSTER is a long-running operation
