# MP023: require-if-not-exists

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
CREATE TABLE/INDEX without IF NOT EXISTS will fail if the object already exists, making migrations non-idempotent.

## Why It's Dangerous
Without IF NOT EXISTS, re-running a migration fails with "relation already exists". Idempotent migrations are safer for retry and rollback scenarios, and required by many deployment pipelines.

## Bad Example
```sql
CREATE TABLE users (
  id bigint PRIMARY KEY,
  email text
);

CREATE INDEX idx_users_email ON users (email);
```

## Good Example
```sql
CREATE TABLE IF NOT EXISTS users (
  id bigint PRIMARY KEY,
  email text
);

CREATE INDEX IF NOT EXISTS idx_users_email ON users (email);
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP023: false           # disable
  MP023:
    severity: critical   # change severity
```

## Related Rules
- [MP001](./MP001.md) require-concurrent-index-creation -- indexes should use CONCURRENTLY
- [MP042](./MP042.md) require-index-name -- indexes should have explicit names
- [MP045](./MP045.md) require-primary-key -- tables should have a primary key
