# MP026: ban-drop-table

**Severity**: Critical | **Auto-fixable**: No | **Tier**: Free

## What It Detects
DROP TABLE permanently removes the table and all its data. Use with extreme caution in production.

## Why It's Dangerous
DROP TABLE is irreversible -- it permanently deletes the table, all rows, indexes, constraints, triggers, and policies. In production, this means instant data loss. Prefer renaming the table first, keeping it as a backup, then dropping later after confirming no dependencies.

## Bad Example
```sql
DROP TABLE legacy_users;
```

## Good Example
```sql
-- Step 1: Rename the table (keeps data as backup)
ALTER TABLE legacy_users RENAME TO legacy_users_deprecated;

-- Step 2: After confirming no application depends on it, drop later
-- DROP TABLE legacy_users_deprecated;
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP026: false           # disable
  MP026:
    severity: warning    # change severity
```

## Related Rules
- [MP022](./MP022.md) no-drop-cascade -- CASCADE silently drops dependent objects
- [MP034](./MP034.md) ban-drop-database -- DROP DATABASE is even more destructive
- [MP017](./MP017.md) no-drop-column -- DROP COLUMN also needs careful coordination
