# MP024: no-enum-value-removal

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
DROP TYPE destroys the enum and all columns using it. PostgreSQL has no ALTER TYPE ... DROP VALUE -- removing enum values requires a full type recreation.

## Why It's Dangerous
Dropping an enum type destroys the type and fails if any columns reference it. PostgreSQL cannot remove individual enum values -- you must recreate the type and migrate all columns, which requires an ACCESS EXCLUSIVE lock per table.

## Bad Example
```sql
DROP TYPE status_type;
```

## Good Example
```sql
-- Safe enum recreation pattern:
-- 1. Create the new type:
CREATE TYPE status_type_new AS ENUM ('active', 'inactive');

-- 2. Migrate columns (ACCESS EXCLUSIVE lock, rewrites table):
ALTER TABLE users ALTER COLUMN status
  TYPE status_type_new USING status::text::status_type_new;

-- 3. Drop the old type:
DROP TYPE status_type;

-- 4. Rename:
ALTER TYPE status_type_new RENAME TO status_type;
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP024: false           # disable
  MP024:
    severity: critical   # change severity
```

## Related Rules
- [MP012](./MP012.md) no-enum-add-value-in-transaction -- adding enum values in transactions
- [MP007](./MP007.md) no-column-type-change -- type changes require table rewrites
- [MP022](./MP022.md) no-drop-cascade -- CASCADE can silently destroy dependents
