# MP028: no-rename-table

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
Renaming a table breaks all application queries, views, and foreign keys referencing the old name.

## Why It's Dangerous
Renaming a table takes an ACCESS EXCLUSIVE lock and instantly breaks every application query, view, function, foreign key, and trigger referencing the old name. Unlike renaming a column, there is no pg_attribute fallback. Use the create-copy-swap pattern for zero-downtime renames.

## Bad Example
```sql
ALTER TABLE users RENAME TO customers;
```

## Good Example
```sql
-- Step 1: Create a view with the old name pointing to the new table
CREATE VIEW users AS SELECT * FROM customers;

-- Step 2: Update application code to use "customers"
-- Step 3: Drop the compatibility view after all code is updated
-- DROP VIEW users;
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP028: false           # disable
  MP028:
    severity: critical   # change severity
```

## Related Rules
- [MP010](./MP010.md) no-rename-column -- renaming columns has similar risks
- [MP026](./MP026.md) ban-drop-table -- dropping tables is irreversible
- [MP017](./MP017.md) no-drop-column -- dropping columns breaks running code
