# MP022: no-drop-cascade

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
DROP ... CASCADE silently drops all dependent objects (views, foreign keys, policies). Always drop dependents explicitly.

## Why It's Dangerous
CASCADE silently drops all dependent objects -- views, foreign keys, policies, and triggers -- without listing them. You may unintentionally destroy critical production objects that other services depend on.

## Bad Example
```sql
DROP TABLE orders CASCADE;
```

## Good Example
```sql
-- First check dependent objects:
SELECT deptype, classid::regclass, objid, objsubid
  FROM pg_depend
  WHERE refobjid = 'orders'::regclass;

-- Then drop dependents explicitly before the target:
DROP VIEW order_summary;
DROP TABLE orders;
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP022: false           # disable
  MP022:
    severity: critical   # change severity
```

## Related Rules
- [MP026](./MP026.md) ban-drop-table -- DROP TABLE itself is dangerous
- [MP035](./MP035.md) ban-drop-schema -- DROP SCHEMA CASCADE is similar
- [MP036](./MP036.md) ban-truncate-cascade -- TRUNCATE CASCADE destroys related data
