# MP032: ban-cluster

**Severity**: Critical | **Auto-fixable**: No | **Tier**: Free

## What It Detects
CLUSTER rewrites the entire table under ACCESS EXCLUSIVE lock to match an index ordering. Use pg_repack instead.

## Why It's Dangerous
CLUSTER physically reorders all rows in the table to match an index, requiring a full table rewrite under ACCESS EXCLUSIVE lock. On large tables this can take hours, blocking all reads and writes. Use pg_repack for online table reorganization without blocking.

## Bad Example
```sql
CLUSTER users USING idx_users_created_at;
```

## Good Example
```sql
-- Use pg_repack for online table reorganization (no ACCESS EXCLUSIVE lock):
-- Install: CREATE EXTENSION pg_repack;
-- Run: pg_repack --table users --order-by idx_users_created_at --no-superuser-check
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP032: false           # disable
  MP032:
    severity: warning    # change severity
```

## Related Rules
- [MP006](./MP006.md) no-vacuum-full -- VACUUM FULL also rewrites the entire table
- [MP020](./MP020.md) require-statement-timeout -- long-running operations need timeouts
- [MP004](./MP004.md) require-lock-timeout -- set lock_timeout before DDL
