# MP039: prefer-identity-over-serial

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
SERIAL/BIGSERIAL creates an implicit sequence with ownership quirks. Use GENERATED ALWAYS AS IDENTITY (PG 10+) instead.

## Why It's Dangerous
SERIAL is a legacy shorthand that creates a separate sequence with confusing ownership semantics -- dropping the column does not drop the sequence, and permissions are not automatically granted. GENERATED ALWAYS AS IDENTITY (PG 10+) is SQL-standard, has cleaner ownership, and is the recommended approach.

## Bad Example
```sql
CREATE TABLE users (
  id SERIAL PRIMARY KEY,
  name TEXT
);
```

## Good Example
```sql
CREATE TABLE users (
  id BIGINT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
  name TEXT
);
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP039: false           # disable
  MP039:
    severity: critical   # change severity
```

## Related Rules
- [MP015](./MP015.md) no-add-column-serial -- ADD COLUMN with SERIAL has similar issues
- [MP038](./MP038.md) prefer-bigint-over-int -- use BIGINT for key columns
- [MP045](./MP045.md) require-primary-key -- every table needs a primary key
