# MP035: ban-drop-schema

**Severity**: Critical | **Auto-fixable**: No | **Tier**: Free

## What It Detects
DROP SCHEMA permanently removes the schema and potentially all objects within it. Use with extreme caution.

## Why It's Dangerous
DROP SCHEMA removes the schema and -- with CASCADE -- all tables, views, functions, and types it contains. Even without CASCADE it takes an ACCESS EXCLUSIVE lock. Dropped schemas cannot be recovered without a backup restore.

## Bad Example
```sql
DROP SCHEMA analytics CASCADE;
```

## Good Example
```sql
-- Verify the schema is empty and unused before dropping:
SELECT * FROM information_schema.tables WHERE table_schema = 'analytics';
SELECT * FROM information_schema.routines WHERE routine_schema = 'analytics';

-- Only drop if confirmed empty:
DROP SCHEMA analytics;
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP035: false           # disable
  MP035:
    severity: warning    # change severity
```

## Related Rules
- [MP034](./MP034.md) ban-drop-database -- DROP DATABASE is even more destructive
- [MP026](./MP026.md) ban-drop-table -- DROP TABLE on individual tables
- [MP022](./MP022.md) no-drop-cascade -- CASCADE silently drops dependent objects
