# MP034: ban-drop-database

**Severity**: Critical | **Auto-fixable**: No | **Tier**: Free

## What It Detects
DROP DATABASE permanently destroys the entire database. This should never appear in a migration file.

## Why It's Dangerous
DROP DATABASE permanently and irreversibly destroys the entire database including all tables, data, indexes, and extensions. If this statement appears in a migration file, it almost certainly indicates an error. There is no undo.

## Bad Example
```sql
DROP DATABASE production;
```

## Good Example
```sql
-- DROP DATABASE should never be in a migration file.
-- If you need to reset a database, use a separate administrative script.
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP034: false           # disable
  MP034:
    severity: warning    # change severity
```

## Related Rules
- [MP035](./MP035.md) ban-drop-schema -- DROP SCHEMA is similarly destructive
- [MP026](./MP026.md) ban-drop-table -- DROP TABLE destroys individual tables
- [MP022](./MP022.md) no-drop-cascade -- CASCADE amplifies destructive operations
