# MP038: prefer-bigint-over-int

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
Primary key and foreign key columns using INT (4 bytes, max 2.1B) can overflow on high-traffic tables. Use BIGINT (8 bytes) instead.

## Why It's Dangerous
INT primary keys overflow at ~2.1 billion rows. Migrating from INT to BIGINT on a large table requires a full table rewrite under ACCESS EXCLUSIVE lock, which can take hours. Starting with BIGINT avoids this expensive migration and costs only 4 extra bytes per row.

## Bad Example
```sql
CREATE TABLE orders (
  id INTEGER PRIMARY KEY,
  user_id INTEGER REFERENCES users(id)
);
```

## Good Example
```sql
CREATE TABLE orders (
  id BIGINT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
  user_id BIGINT REFERENCES users(id)
);
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP038: false           # disable
  MP038:
    severity: critical   # change severity
```

## Related Rules
- [MP039](./MP039.md) prefer-identity-over-serial -- use IDENTITY instead of SERIAL
- [MP045](./MP045.md) require-primary-key -- every table needs a primary key
- [MP007](./MP007.md) no-column-type-change -- INT to BIGINT migration is very expensive
