# MP036: ban-truncate-cascade

**Severity**: Critical | **Auto-fixable**: No | **Tier**: Free

## What It Detects
TRUNCATE CASCADE silently truncates all tables with foreign key references to the target. This can destroy data across many tables.

## Why It's Dangerous
TRUNCATE CASCADE removes all rows not only from the target table but from every table that references it via foreign keys, recursively. This can silently wipe data from dozens of tables. Always truncate specific tables explicitly.

## Bad Example
```sql
TRUNCATE orders CASCADE;
```

## Good Example
```sql
-- Truncate specific tables explicitly in dependency order:
TRUNCATE order_items;
TRUNCATE orders;
-- Or use DELETE with WHERE clauses for safer, auditable data removal.
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP036: false           # disable
  MP036:
    severity: warning    # change severity
```

## Related Rules
- [MP022](./MP022.md) no-drop-cascade -- CASCADE on DROP is equally dangerous
- [MP026](./MP026.md) ban-drop-table -- DROP TABLE is destructive
- [MP034](./MP034.md) ban-drop-database -- database-level destruction
