# MP037: prefer-text-over-varchar

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
VARCHAR(n) offers no performance benefit over TEXT in PostgreSQL and makes future schema changes harder.

## Why It's Dangerous
In PostgreSQL, VARCHAR(n) and TEXT have identical performance -- both use the same varlena storage. VARCHAR(n) only adds a length check constraint that makes future length changes require a table rewrite (on PG < 17) or at minimum a constraint adjustment. Use TEXT with a CHECK constraint if you need length validation.

## Bad Example
```sql
CREATE TABLE users (
  id bigint PRIMARY KEY,
  email VARCHAR(255),
  name VARCHAR(100)
);
```

## Good Example
```sql
CREATE TABLE users (
  id bigint PRIMARY KEY,
  email TEXT,
  name TEXT CHECK (length(name) <= 100)
);
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP037: false           # disable
  MP037:
    severity: critical   # change severity
```

## Related Rules
- [MP041](./MP041.md) ban-char-field -- CHAR(n) is even worse than VARCHAR(n)
- [MP007](./MP007.md) no-column-type-change -- changing VARCHAR length requires rewrites
- [MP040](./MP040.md) prefer-timestamptz -- use correct types from the start
