# MP040: prefer-timestamptz

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
TIMESTAMP WITHOUT TIME ZONE loses timezone information, causing bugs when servers or users span time zones. Use TIMESTAMPTZ instead.

## Why It's Dangerous
TIMESTAMP WITHOUT TIME ZONE stores the literal wall-clock time with no timezone context. When servers move between zones, or users are in different timezones, this causes silent data corruption. TIMESTAMPTZ stores instants in UTC and converts on display, which is almost always what you want.

## Bad Example
```sql
CREATE TABLE events (
  id bigint PRIMARY KEY,
  created_at TIMESTAMP,
  updated_at TIMESTAMP
);
```

## Good Example
```sql
CREATE TABLE events (
  id bigint PRIMARY KEY,
  created_at TIMESTAMPTZ,
  updated_at TIMESTAMPTZ
);
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP040: false           # disable
  MP040:
    severity: critical   # change severity
```

## Related Rules
- [MP037](./MP037.md) prefer-text-over-varchar -- choose correct types from the start
- [MP041](./MP041.md) ban-char-field -- avoid problematic types
- [MP007](./MP007.md) no-column-type-change -- changing types later is very expensive
