# MP049: require-partition-key-in-pk

**Severity**: Critical | **Auto-fixable**: No | **Tier**: Free

## What It Detects
Partitioned tables where the primary key does not include all partition key columns.

## Why It's Dangerous
PostgreSQL requires that the primary key (and all unique constraints) on a partitioned table include all partition key columns. If omitted, the CREATE TABLE will fail at runtime. This ensures uniqueness can be enforced per-partition.

## Bad Example
```sql
CREATE TABLE events (
  id BIGINT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
  tenant_id bigint NOT NULL,
  event_type text,
  created_at timestamptz
) PARTITION BY HASH (tenant_id);
```

## Good Example
```sql
CREATE TABLE events (
  id BIGINT GENERATED ALWAYS AS IDENTITY,
  tenant_id bigint NOT NULL,
  event_type text,
  created_at timestamptz,
  PRIMARY KEY (id, tenant_id)
) PARTITION BY HASH (tenant_id);
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP049: false           # disable
  MP049:
    severity: warning    # change severity
```

## Related Rules
- [MP045](./MP045.md) require-primary-key -- every table should have a primary key
- [MP038](./MP038.md) prefer-bigint-over-int -- use BIGINT for primary keys
- [MP023](./MP023.md) require-if-not-exists -- CREATE TABLE should use IF NOT EXISTS
