# MP042: require-index-name

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
CREATE INDEX without an explicit name generates auto-names that are hard to reference in future migrations and monitoring.

## Why It's Dangerous
PostgreSQL auto-generates index names like "users_email_idx" but the naming is fragile and can collide. Explicit index names make future migrations clearer (DROP INDEX, REINDEX), improve monitoring (pg_stat_user_indexes), and are required for UNIQUE ... USING INDEX patterns.

## Bad Example
```sql
CREATE INDEX ON users (email);
```

## Good Example
```sql
CREATE INDEX idx_users_email ON users (email);
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP042: false           # disable
  MP042:
    severity: critical   # change severity
```

## Related Rules
- [MP001](./MP001.md) require-concurrent-index-creation -- indexes should use CONCURRENTLY
- [MP023](./MP023.md) require-if-not-exists -- indexes should use IF NOT EXISTS
- [MP027](./MP027.md) disallowed-unique-constraint -- UNIQUE constraints need named indexes
