# MP041: ban-char-field

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
CHAR(n) pads values with spaces, wasting storage and causing subtle bugs in comparisons. Use TEXT or VARCHAR instead.

## Why It's Dangerous
CHAR(n) blank-pads values to the specified length, wasting storage and causing confusing behavior in string comparisons and LIKE queries. It offers no performance advantage over TEXT in PostgreSQL. Use TEXT (or VARCHAR if you must have a length limit).

## Bad Example
```sql
CREATE TABLE countries (
  code CHAR(2) PRIMARY KEY,
  name CHAR(100)
);
```

## Good Example
```sql
CREATE TABLE countries (
  code TEXT PRIMARY KEY CHECK (length(code) = 2),
  name TEXT
);
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP041: false           # disable
  MP041:
    severity: critical   # change severity
```

## Related Rules
- [MP037](./MP037.md) prefer-text-over-varchar -- VARCHAR is better than CHAR but TEXT is best
- [MP040](./MP040.md) prefer-timestamptz -- use correct types from the start
- [MP007](./MP007.md) no-column-type-change -- changing types later is expensive
