# MP047: ban-set-logged-unlogged

**Severity**: Critical | **Auto-fixable**: No | **Tier**: Free

## What It Detects
SET LOGGED/UNLOGGED rewrites the entire table under ACCESS EXCLUSIVE lock. This blocks all reads and writes for the duration.

## Why It's Dangerous
Changing a table between LOGGED and UNLOGGED requires a full table rewrite while holding ACCESS EXCLUSIVE lock. On large tables this can take hours, blocking all queries. UNLOGGED tables also do not survive crash recovery -- all data is lost on restart.

## Bad Example
```sql
ALTER TABLE events SET UNLOGGED;
```

## Good Example
```sql
-- Changing LOGGED/UNLOGGED requires a full table rewrite.
-- Consider performing this during a maintenance window.
-- For LOGGED to UNLOGGED: ensure you have backups, data will be lost on crash.
-- For UNLOGGED to LOGGED: consider creating a new LOGGED table and migrating data.
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP047: false           # disable
  MP047:
    severity: warning    # change severity
```

## Related Rules
- [MP006](./MP006.md) no-vacuum-full -- VACUUM FULL also rewrites the entire table
- [MP032](./MP032.md) ban-cluster -- CLUSTER also rewrites the entire table
- [MP020](./MP020.md) require-statement-timeout -- long-running operations need timeouts
