# MP045: require-primary-key

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
Tables without a primary key cannot be efficiently replicated and make row identification ambiguous.

## Why It's Dangerous
Tables without a primary key cannot use logical replication (a requirement for zero-downtime upgrades), make UPDATE/DELETE operations ambiguous, prevent efficient foreign key references, and break many ORMs. Add a primary key to every table.

## Bad Example
```sql
CREATE TABLE events (
  user_id bigint,
  event_type text,
  created_at timestamptz
);
```

## Good Example
```sql
CREATE TABLE events (
  id BIGINT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
  user_id bigint,
  event_type text,
  created_at timestamptz
);
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP045: false           # disable
  MP045:
    severity: critical   # change severity
```

## Related Rules
- [MP038](./MP038.md) prefer-bigint-over-int -- use BIGINT for primary keys
- [MP039](./MP039.md) prefer-identity-over-serial -- use IDENTITY for auto-increment PKs
- [MP023](./MP023.md) require-if-not-exists -- CREATE TABLE should use IF NOT EXISTS
