# MP046: require-concurrent-detach-partition

**Severity**: Critical | **Auto-fixable**: No | **Tier**: Free

## What It Detects
DETACH PARTITION without CONCURRENTLY takes ACCESS EXCLUSIVE lock on the parent, blocking all queries. Use CONCURRENTLY on PG 14+.

## Why It's Dangerous
DETACH PARTITION without CONCURRENTLY takes an ACCESS EXCLUSIVE lock on the parent table, blocking all reads and writes until the operation completes. DETACH PARTITION CONCURRENTLY (PG 14+) uses a two-phase approach that only briefly locks the parent.

## Bad Example
```sql
ALTER TABLE events DETACH PARTITION events_2023;
```

## Good Example
```sql
-- Use CONCURRENTLY to avoid blocking (PG 14+):
ALTER TABLE events DETACH PARTITION events_2023 CONCURRENTLY;
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP046: false           # disable
  MP046:
    severity: warning    # change severity
```

## Related Rules
- [MP001](./MP001.md) require-concurrent-index-creation -- CONCURRENTLY for index creation
- [MP008](./MP008.md) no-multi-ddl-transaction -- DDL should not be wrapped in transactions
- [MP004](./MP004.md) require-lock-timeout -- set lock_timeout before DDL
