# MP044: no-data-loss-type-narrowing

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
Changing a column to a narrower type can cause data loss or errors if existing values exceed the new type range.

## Why It's Dangerous
Narrowing a column type (e.g., BIGINT to INT, TEXT to VARCHAR(50)) will fail if any existing row has a value that does not fit the new type. Even if it succeeds today, future inserts may fail unexpectedly. The change also requires a full table rewrite under ACCESS EXCLUSIVE lock.

## Bad Example
```sql
ALTER TABLE events ALTER COLUMN counter TYPE integer;
-- If counter was BIGINT and any value exceeds 2.1B, this fails
```

## Good Example
```sql
-- Verify no data exceeds the new type range before changing:
SELECT COUNT(*) FROM events WHERE counter > 2147483647;

-- Consider adding a CHECK constraint instead of narrowing the type:
ALTER TABLE events ADD CONSTRAINT events_counter_range
  CHECK (counter <= 2147483647) NOT VALID;
ALTER TABLE events VALIDATE CONSTRAINT events_counter_range;
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP044: false           # disable
  MP044:
    severity: critical   # change severity
```

## Related Rules
- [MP007](./MP007.md) no-column-type-change -- all type changes require table rewrites
- [MP038](./MP038.md) prefer-bigint-over-int -- start with BIGINT to avoid future narrowing
- [MP030](./MP030.md) require-not-valid-check -- use CHECK constraints for validation
