# MP043: ban-domain-constraint

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
Adding or modifying domain constraints validates against ALL columns using that domain, potentially scanning many tables.

## Why It's Dangerous
Domain constraints are validated against every column in every table that uses the domain type. Adding a CHECK constraint to a domain can trigger full table scans across many tables simultaneously. Use CHECK constraints on individual columns instead.

## Bad Example
```sql
CREATE DOMAIN email_address AS TEXT
  CHECK (VALUE ~ '^[^@]+@[^@]+\.[^@]+$');

-- Or modifying an existing domain:
ALTER DOMAIN email_address ADD CONSTRAINT valid_email
  CHECK (length(VALUE) <= 255);
```

## Good Example
```sql
-- Use CHECK constraints on individual columns instead of domain constraints:
CREATE TABLE users (
  email TEXT CHECK (email ~ '^[^@]+@[^@]+\.[^@]+$')
);
-- This gives you more control over validation and avoids cross-table scans.
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP043: false           # disable
  MP043:
    severity: critical   # change severity
```

## Related Rules
- [MP030](./MP030.md) require-not-valid-check -- CHECK constraints should use NOT VALID
- [MP037](./MP037.md) prefer-text-over-varchar -- use TEXT with CHECK for validation
- [MP007](./MP007.md) no-column-type-change -- domain changes affect all dependent columns
