# MP052: warn-dependent-objects

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
DROP COLUMN, RENAME COLUMN, or ALTER COLUMN TYPE operations that may silently break views, functions, and triggers that reference the column.

## Why It's Dangerous
Views, functions, and triggers that reference a column will fail at query time — not at migration time — when the column is dropped, renamed, or its type is changed. PostgreSQL does not automatically update these dependent objects. This is the most common cause of post-deployment failures that static analysis cannot fully resolve without a database connection.

## Bad Example
```sql
ALTER TABLE users DROP COLUMN email;
-- Views that SELECT email will fail at query time, not migration time
```

## Good Example
```sql
-- Check for dependent objects first:
SELECT dependent_ns.nspname || '.' || dependent_view.relname AS dependent_object,
       dependent_view.relkind
FROM pg_depend
JOIN pg_rewrite ON pg_depend.objid = pg_rewrite.oid
JOIN pg_class AS dependent_view ON pg_rewrite.ev_class = dependent_view.oid
JOIN pg_namespace AS dependent_ns ON dependent_view.relnamespace = dependent_ns.oid
JOIN pg_class AS source_table ON pg_depend.refobjid = source_table.oid
JOIN pg_attribute ON pg_depend.refobjid = pg_attribute.attrelid
  AND pg_depend.refobjsubid = pg_attribute.attnum
WHERE source_table.relname = 'users' AND pg_attribute.attname = 'email';

-- Only drop after confirming no dependents:
ALTER TABLE users DROP COLUMN email;
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP052: false           # disable
  MP052:
    severity: critical   # change severity
```

## Related Rules
- [MP017](./MP017.md) no-drop-column -- DROP COLUMN acquires ACCESS EXCLUSIVE lock
- [MP010](./MP010.md) no-rename-column -- renaming columns breaks application queries
- [MP007](./MP007.md) no-column-type-change -- changing column types requires table rewrite
