# MP058: multi-alter-table-same-table

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
Multiple separate ALTER TABLE statements targeting the same table within the same migration file.

## Why It's Dangerous
Each ALTER TABLE acquires ACCESS EXCLUSIVE lock independently. Multiple separate statements mean multiple lock/unlock cycles, each going through the lock queue. Combining subcommands into a single ALTER TABLE reduces the blocking window from N separate lock cycles to one.

## Bad Example
```sql
ALTER TABLE users ADD COLUMN bio text;
ALTER TABLE users ADD COLUMN avatar text;
ALTER TABLE users ALTER COLUMN bio SET DEFAULT '';
-- 3 separate lock acquisitions
```

## Good Example
```sql
ALTER TABLE users
  ADD COLUMN bio text DEFAULT '',
  ADD COLUMN avatar text;
-- 1 lock acquisition
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP058: false           # disable
  MP058:
    severity: critical   # change severity
```

## Related Rules
- [MP004](./MP004.md) require-lock-timeout -- SET lock_timeout before DDL
- [MP008](./MP008.md) no-multi-ddl-transaction -- avoid multiple DDL in one transaction
