# MP051: require-spatial-index

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
Tables with spatial/geometry columns (geometry, geography, point, line, lseg, box, path, polygon, circle) that do not have a corresponding GIST or SP-GIST index.

## Why It's Dangerous
PostGIS geometry and geography columns need a GIST or SP-GIST index for efficient spatial queries (ST_Contains, ST_DWithin, etc.). Without one, every spatial query triggers a full sequential scan. Adding the index at table creation time avoids the need for a later lock-heavy ALTER.

## Bad Example
```sql
CREATE TABLE locations (
  id BIGINT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
  name text NOT NULL,
  geom geometry(Point, 4326)
);
```

## Good Example
```sql
CREATE TABLE locations (
  id BIGINT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
  name text NOT NULL,
  geom geometry(Point, 4326)
);

CREATE INDEX CONCURRENTLY idx_locations_geom ON locations USING GIST (geom);
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP051: false           # disable
  MP051:
    severity: critical   # change severity
```

## Related Rules
- [MP001](./MP001.md) require-concurrent-index -- CREATE INDEX must use CONCURRENTLY
- [MP042](./MP042.md) require-index-name -- indexes should have explicit names
- [MP050](./MP050.md) prefer-hnsw-over-ivfflat -- use HNSW for vector indexes
