# MP054: alter-type-add-value-in-transaction

**Severity**: Critical | **Auto-fixable**: No | **Tier**: Free

## What It Detects
INSERT or UPDATE statements that reference a newly added enum value in the same transaction as the ALTER TYPE ADD VALUE that created it.

## Why It's Dangerous
On PostgreSQL < 12, ALTER TYPE ADD VALUE cannot run inside a transaction at all. On PG 12+, it can run in a transaction but the new enum value is not visible to other statements in the same transaction — any INSERT or UPDATE referencing the new value will fail with "unsafe use of new value." Prisma, TypeORM, and Alembic frequently generate migrations that trigger this.

## Bad Example
```sql
BEGIN;
ALTER TYPE status ADD VALUE 'archived';
INSERT INTO events (status) VALUES ('archived');
-- Fails: "unsafe use of new value" — the value isn't visible until COMMIT
COMMIT;
```

## Good Example
```sql
-- Transaction 1: add the enum value
ALTER TYPE status ADD VALUE 'archived';

-- Transaction 2 (after the first commits): use the value
INSERT INTO events (status) VALUES ('archived');
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required — split into separate transactions.

## Configuration
```yaml
rules:
  MP054: false           # disable
  MP054:
    severity: warning    # change severity
```

## Related Rules
- [MP012](./MP012.md) no-enum-add-in-transaction -- ALTER TYPE ADD VALUE cannot run in transaction (PG < 12)
- [MP024](./MP024.md) no-enum-value-removal -- DROP TYPE destroys enum and dependent columns
- [MP053](./MP053.md) ban-uncommitted-transaction -- BEGIN without matching COMMIT
