# MP053: ban-uncommitted-transaction

**Severity**: Critical | **Auto-fixable**: No | **Tier**: Free

## What It Detects
Migration files that contain BEGIN without a matching COMMIT or ROLLBACK, leaving a dangling open transaction.

## Why It's Dangerous
A migration with BEGIN but no COMMIT will either fail (if the migration runner auto-commits) or leave an open transaction that holds locks indefinitely. Conversely, a COMMIT without a preceding BEGIN is a no-op but indicates a structural problem. Always match BEGIN with COMMIT or ROLLBACK.

## Bad Example
```sql
BEGIN;
ALTER TABLE users ADD COLUMN bio text;
-- Missing COMMIT â€” this transaction stays open, holding locks
```

## Good Example
```sql
BEGIN;
ALTER TABLE users ADD COLUMN bio text;
COMMIT;
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP053: false           # disable
  MP053:
    severity: warning    # change severity
```

## Related Rules
- [MP008](./MP008.md) no-multi-ddl-transaction -- avoid multiple DDL in one transaction
- [MP025](./MP025.md) ban-concurrent-in-transaction -- CONCURRENTLY cannot run inside a transaction
- [MP054](./MP054.md) alter-type-add-value-in-transaction -- new enum values not visible until COMMIT
