# MP050: prefer-hnsw-over-ivfflat

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
pgvector indexes using the IVFFlat access method instead of HNSW.

## Why It's Dangerous
pgvector IVFFlat indexes need enough representative data at creation time to build good clusters. As data distribution changes, recall degrades and you must REINDEX. HNSW indexes build incrementally, have consistently better recall, and never need retraining. Prefer HNSW unless you have a specific reason to use IVFFlat.

## Bad Example
```sql
CREATE INDEX idx_embeddings ON documents USING ivfflat (embedding vector_cosine_ops)
  WITH (lists = 100);
```

## Good Example
```sql
CREATE INDEX idx_embeddings ON documents USING hnsw (embedding vector_cosine_ops)
  WITH (m = 16, ef_construction = 64);
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP050: false           # disable
  MP050:
    severity: critical   # change severity
```

## Related Rules
- [MP001](./MP001.md) require-concurrent-index -- CREATE INDEX must use CONCURRENTLY
- [MP042](./MP042.md) require-index-name -- indexes should have explicit names
