# MP059: sequence-not-reset-after-data-migration

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
INSERT with explicit integer IDs without a subsequent setval() or ALTER SEQUENCE RESTART to reset the sequence counter.

## Why It's Dangerous
When you seed data or migrate rows with explicit integer IDs, the sequence counter is not automatically updated. The next auto-generated INSERT picks a low ID that already exists, causing "duplicate key violates unique constraint." This is one of the most common post-migration production errors.

## Bad Example
```sql
INSERT INTO users (id, name) VALUES (1, 'Alice');
INSERT INTO users (id, name) VALUES (2, 'Bob');
-- Next auto-generated INSERT gets id=1 â†’ duplicate key error!
```

## Good Example
```sql
INSERT INTO users (id, name) VALUES (1, 'Alice');
INSERT INTO users (id, name) VALUES (2, 'Bob');
SELECT setval(pg_get_serial_sequence('users', 'id'), COALESCE(MAX(id), 1)) FROM users;
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP059: false           # disable
  MP059:
    severity: critical   # change severity
```

## Related Rules
- [MP015](./MP015.md) no-add-column-serial -- SERIAL creates implicit sequence
- [MP039](./MP039.md) prefer-identity-over-serial -- GENERATED AS IDENTITY is safer
