# MP057: rls-enabled-without-policy

**Severity**: Critical | **Auto-fixable**: No | **Tier**: Free

## What It Detects
ENABLE ROW LEVEL SECURITY without a matching CREATE POLICY in the same migration file.

## Why It's Dangerous
When RLS is enabled with zero policies, the default behavior is a complete deny â€” all queries from non-superuser roles return zero rows. No error is raised. SELECT * FROM table returns empty. Supabase documents this as the leading cause of data exposure/lockout incidents on their platform.

## Bad Example
```sql
ALTER TABLE users ENABLE ROW LEVEL SECURITY;
-- All non-superuser queries now return 0 rows!
```

## Good Example
```sql
ALTER TABLE users ENABLE ROW LEVEL SECURITY;
CREATE POLICY users_select ON users FOR SELECT USING (true);
CREATE POLICY users_insert ON users FOR INSERT WITH CHECK (auth.uid() = user_id);
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP057: false           # disable
  MP057:
    severity: warning    # change severity
```

## Related Rules
- [MP026](./MP026.md) ban-drop-table -- dropping tables is destructive
- [MP022](./MP022.md) no-drop-cascade -- CASCADE silently drops dependent objects
