# MP056: gin-index-on-jsonb-without-expression

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
A plain GIN index on a JSONB column without an expression or jsonb_path_ops operator class.

## Why It's Dangerous
A GIN index with default jsonb_ops only supports containment operators (@>, ?, ?|, ?&). The common ->> extraction operator used by ORMs (Prisma, TypeORM, Knex) will still do a sequential scan. Use an expression B-tree index for ->> queries, or jsonb_path_ops if you only use @> containment queries.

## Bad Example
```sql
CREATE INDEX idx_events_data ON events USING GIN (data);
-- Useless for: WHERE data->>'status' = 'active'
```

## Good Example
```sql
-- For ->> queries:
CREATE INDEX idx_events_status ON events ((data->>'status'));

-- For @> containment queries:
CREATE INDEX idx_events_data ON events USING GIN (data jsonb_path_ops);
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP056: false           # disable
  MP056:
    severity: critical   # change severity
```

## Related Rules
- [MP042](./MP042.md) require-index-name -- indexes should have explicit names
- [MP001](./MP001.md) require-concurrent-index -- CREATE INDEX must use CONCURRENTLY
