# MP055: drop-pk-replica-identity-break

**Severity**: Critical | **Auto-fixable**: No | **Tier**: Free

## What It Detects
Dropping a primary key constraint without setting REPLICA IDENTITY on the table.

## Why It's Dangerous
The default replica identity IS the primary key. When you drop a PK without setting REPLICA IDENTITY FULL or USING INDEX, all subsequent UPDATE and DELETE operations fail on logical replication subscribers (Supabase, Neon, AWS RDS read replicas, Debezium CDC). The publisher succeeds but the subscriber silently errors.

## Bad Example
```sql
ALTER TABLE users DROP CONSTRAINT users_pkey;
-- Logical replication breaks: cannot delete/update without replica identity
```

## Good Example
```sql
ALTER TABLE users REPLICA IDENTITY FULL;
ALTER TABLE users DROP CONSTRAINT users_pkey;
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP055: false           # disable
  MP055:
    severity: warning    # change severity
```

## Related Rules
- [MP045](./MP045.md) require-primary-key -- tables should always have a PK
- [MP017](./MP017.md) no-drop-column -- dropping columns requires ACCESS EXCLUSIVE lock
