# MP065: ban-lock-table

**Severity**: Critical | **Auto-fixable**: No | **Tier**: Free

## What It Detects
Explicit LOCK TABLE statements in migration files.

## Why It's Dangerous
LOCK TABLE acquires an explicit lock that can block reads and writes depending on the mode. PostgreSQL DDL statements automatically acquire the correct lock â€” explicit LOCK TABLE is rarely needed and often indicates a flawed migration strategy. High lock modes (EXCLUSIVE, ACCESS EXCLUSIVE) block all other operations. Even weaker modes can cause deadlocks when combined with other concurrent operations.

## Bad Example
```sql
LOCK TABLE users IN ACCESS EXCLUSIVE MODE;
-- Blocks ALL operations (reads and writes) on the table
-- Holds the lock until the transaction ends

LOCK TABLE orders IN SHARE MODE;
-- Blocks writes but allows reads
-- Can cause deadlocks with concurrent transactions
```

## Good Example
```sql
-- Let PostgreSQL acquire locks automatically via DDL
-- No explicit LOCK TABLE needed
ALTER TABLE users ADD COLUMN email TEXT;

-- If you need to coordinate operations, use advisory locks instead
SELECT pg_advisory_lock(12345);
-- ... perform operations ...
SELECT pg_advisory_unlock(12345);
```

## Auto-fix
This rule is not auto-fixable. Remove the explicit LOCK TABLE statement and let PostgreSQL acquire locks automatically through DDL operations.

## Configuration
```yaml
rules:
  MP065: false           # disable
  MP065:
    severity: warning    # downgrade from critical to warning
```

## Links
- [PostgreSQL LOCK TABLE](https://www.postgresql.org/docs/current/sql-lock.html)
- [PostgreSQL Lock Types](https://www.postgresql.org/docs/current/explicit-locking.html)
