# MP061: suboptimal-column-order

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
CREATE TABLE statements where variable-length columns (text, varchar, jsonb, bytea) appear before fixed-size columns (int, bigint, timestamp, uuid).

## Why It's Dangerous
PostgreSQL stores tuple data in column declaration order. Fixed-size types have specific alignment requirements (2, 4, or 8 bytes). When a variable-length column sits between two fixed-size columns, the second fixed-size column may need padding bytes to meet its alignment requirement. On tables with millions of rows, this wastes 4-16 bytes per row — potentially gigabytes of wasted storage.

## Bad Example
```sql
CREATE TABLE users (
  name TEXT,
  bio TEXT,
  id INTEGER,
  age INTEGER
);
-- Variable-length columns (name, bio) before fixed-size (id, age)
-- wastes alignment padding on every row
```

## Good Example
```sql
CREATE TABLE users (
  id INTEGER,
  age INTEGER,
  name TEXT,
  bio TEXT
);
-- Fixed-size columns first, variable-length last — minimal padding
```

## Auto-fix
This rule is not auto-fixable. Reorder columns manually during table creation.

## Configuration
```yaml
rules:
  MP061: false           # disable
  MP061:
    severity: critical   # upgrade from warning to critical
```

## Links
- [PostgreSQL Physical Storage](https://www.postgresql.org/docs/current/storage-page-layout.html)
- [Column Tetris](https://www.2ndquadrant.com/en/blog/on-rocks-and-sand/)
