# MP060: alter-type-rename-value

**Severity**: Critical | **Auto-fixable**: No | **Tier**: Free

## What It Detects
ALTER TYPE ... RENAME VALUE statements that rename enum values.

## Why It's Dangerous
RENAME VALUE modifies the pg_enum catalog entry in-place. Logical replication does not replicate catalog changes, so subscribers (Neon branches, Supabase read replicas, RDS logical replicas, Debezium CDC) retain the old value name. Any row containing the renamed value that arrives via replication will fail to decode.

## Bad Example
```sql
ALTER TYPE status RENAME VALUE 'active' TO 'enabled';
-- Logical replication subscribers still have 'active', not 'enabled'
```

## Good Example
```sql
-- Add new value, migrate data, then handle old value:
ALTER TYPE status ADD VALUE 'enabled';
UPDATE events SET status = 'enabled' WHERE status = 'active';
```

## Auto-fix
This rule is not auto-fixable. Manual intervention required.

## Configuration
```yaml
rules:
  MP060: false           # disable
  MP060:
    severity: warning    # change severity
```

## Related Rules
- [MP012](./MP012.md) no-enum-add-in-transaction -- ALTER TYPE ADD VALUE cannot run in transaction
- [MP024](./MP024.md) no-enum-value-removal -- dropping enum values is destructive
- [MP054](./MP054.md) alter-type-add-value-in-transaction -- new enum value not visible until COMMIT
