# MP063: warn-do-block-ddl

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
DO $$ ... $$ blocks that contain DDL keywords (ALTER TABLE, CREATE INDEX, DROP TABLE, etc.) in the PL/pgSQL body.

## Why It's Dangerous
PL/pgSQL DO blocks execute arbitrary code that cannot be analyzed by SQL linters. DDL inside DO blocks acquires the same locks as direct SQL statements, but the operations are invisible to static analysis tools. This means dangerous patterns (non-concurrent index creation, column type changes, table drops) can slip through CI checks undetected.

## Bad Example
```sql
DO $$
BEGIN
  ALTER TABLE users ADD COLUMN age INTEGER;
  CREATE INDEX idx_users_age ON users (age);
END;
$$;
-- DDL inside DO block bypasses all static analysis
-- CREATE INDEX without CONCURRENTLY won't be flagged
```

## Good Example
```sql
-- Extract DDL into direct SQL statements
ALTER TABLE users ADD COLUMN age INTEGER;
CREATE INDEX CONCURRENTLY idx_users_age ON users (age);
-- Each statement is individually analyzed for lock safety
```

## Auto-fix
This rule is not auto-fixable. Manually extract DDL from DO blocks into direct SQL statements.

## Configuration
```yaml
rules:
  MP063: false           # disable
  MP063:
    severity: critical   # upgrade from warning to critical
```

## Links
- [PostgreSQL DO Statement](https://www.postgresql.org/docs/current/sql-do.html)
- [PL/pgSQL Overview](https://www.postgresql.org/docs/current/plpgsql.html)
