# MP064: ban-disable-trigger

**Severity**: Critical | **Auto-fixable**: No | **Tier**: Free

## What It Detects
ALTER TABLE ... DISABLE TRIGGER ALL or ALTER TABLE ... DISABLE TRIGGER USER, which disables all triggers on a table.

## Why It's Dangerous
DISABLE TRIGGER ALL/USER turns off all triggers on the table. This breaks logical replication (which uses triggers internally), disables audit logging triggers, and bypasses foreign key enforcement. If the session crashes before re-enabling triggers, they remain disabled permanently. On systems using logical replication (Supabase, Neon, RDS), this can silently break data replication.

## Bad Example
```sql
ALTER TABLE users DISABLE TRIGGER ALL;
-- Breaks logical replication
-- Bypasses audit triggers
-- Skips FK enforcement
-- If session crashes, triggers stay disabled
```

## Good Example
```sql
-- Disable only a specific trigger temporarily
ALTER TABLE users DISABLE TRIGGER my_audit_trigger;
-- ... perform operation ...
ALTER TABLE users ENABLE TRIGGER my_audit_trigger;

-- Or restructure the migration to avoid trigger disabling entirely
```

## Auto-fix
This rule is not auto-fixable. Restructure the migration to avoid disabling triggers, or use targeted trigger disabling for specific named triggers only.

## Configuration
```yaml
rules:
  MP064: false           # disable
  MP064:
    severity: warning    # downgrade from critical to warning
```

## Links
- [PostgreSQL ALTER TABLE](https://www.postgresql.org/docs/current/sql-altertable.html)
- [Logical Replication and Triggers](https://www.postgresql.org/docs/current/logical-replication.html)
