# MP068: warn-integer-pk-capacity

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
CREATE SEQUENCE with integer type, which has a maximum value of approximately 2.1 billion.

## Why It's Dangerous
A sequence using the integer type (4 bytes) maxes out at 2,147,483,647. For tables with high insert rates, this limit can be reached in months or years, causing INSERT failures with "nextval: reached maximum value of sequence" errors. Changing the sequence type on a live table requires an ACCESS EXCLUSIVE lock and a full table rewrite. Using bigint (8 bytes) from the start avoids this entirely, with a maximum of 9.2 quintillion â€” effectively unlimited for any workload.

## Bad Example
```sql
CREATE SEQUENCE user_id_seq AS integer;
-- Max value ~2.1 billion
-- Will fail silently when exhausted
-- Changing later requires table rewrite
```

## Good Example
```sql
CREATE SEQUENCE user_id_seq AS bigint;
-- Max value ~9.2 quintillion
-- Effectively unlimited for any workload
-- No migration needed later
```

## Auto-fix
This rule is not auto-fixable. Change the sequence type from integer to bigint. If the sequence is already in use, use ALTER SEQUENCE ... AS bigint (requires PostgreSQL 10+).

## Configuration
```yaml
rules:
  MP068: false           # disable
  MP068:
    severity: critical   # upgrade from warning to critical
```

## Links
- [PostgreSQL CREATE SEQUENCE](https://www.postgresql.org/docs/current/sql-createsequence.html)
- [PostgreSQL Numeric Types](https://www.postgresql.org/docs/current/datatype-numeric.html)
