# MP066: warn-autovacuum-disabled

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
CREATE TABLE ... WITH (autovacuum_enabled = false) or ALTER TABLE ... SET (autovacuum_enabled = false), which disables autovacuum on the table.

## Why It's Dangerous
Autovacuum performs two critical maintenance tasks: reclaiming dead tuples to prevent table bloat, and advancing the transaction ID to prevent wraparound. Without autovacuum, dead tuples accumulate causing table bloat and degraded query performance. Worse, if the transaction ID counter wraps around, PostgreSQL must shut down to prevent data corruption â€” requiring manual VACUUM to recover.

## Bad Example
```sql
CREATE TABLE staging_data (id INT, data JSONB)
  WITH (autovacuum_enabled = false);
-- Dead tuples accumulate
-- Table bloat degrades performance
-- Risk of transaction ID wraparound

ALTER TABLE users SET (autovacuum_enabled = false);
-- Same risks on an existing production table
```

## Good Example
```sql
-- If temporarily disabling for bulk load, re-enable immediately after
CREATE TABLE staging_data (id INT, data JSONB)
  WITH (autovacuum_enabled = false);

-- After bulk load completes:
ALTER TABLE staging_data SET (autovacuum_enabled = true);
ANALYZE staging_data;

-- Or tune autovacuum instead of disabling it
ALTER TABLE high_churn_table SET (
  autovacuum_vacuum_scale_factor = 0.01,
  autovacuum_analyze_scale_factor = 0.005
);
```

## Auto-fix
This rule is not auto-fixable. Either remove the autovacuum_enabled = false option, or ensure you have a plan to re-enable autovacuum after the bulk operation completes.

## Configuration
```yaml
rules:
  MP066: false           # disable
  MP066:
    severity: critical   # upgrade from warning to critical
```

## Links
- [PostgreSQL Autovacuum](https://www.postgresql.org/docs/current/routine-vacuuming.html#AUTOVACUUM)
- [Transaction ID Wraparound](https://www.postgresql.org/docs/current/routine-vacuuming.html#VACUUM-FOR-WRAPAROUND)
