# MP077: prefer-lz4-toast-compression

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
ALTER TABLE ... SET COMPRESSION pglz, which uses the legacy pglz TOAST compression algorithm instead of lz4.

## Why It's Dangerous
The default pglz compression algorithm in PostgreSQL is significantly slower than lz4, which was introduced in PostgreSQL 14. On write-heavy workloads with large TEXT, JSONB, or BYTEA columns, pglz compression becomes a CPU bottleneck. lz4 provides 3-5x faster compression and decompression with comparable compression ratios for most workloads. On PostgreSQL 14+, there is no reason to explicitly set pglz compression — lz4 should be preferred for better performance on both reads and writes.

## Bad Example
```sql
ALTER TABLE users ALTER COLUMN bio SET COMPRESSION pglz;
-- Slower compression and decompression
-- CPU bottleneck on write-heavy workloads
-- No benefit over lz4 for most data patterns
```

## Good Example
```sql
ALTER TABLE users ALTER COLUMN bio SET COMPRESSION lz4;
-- 3-5x faster compression/decompression
-- Lower CPU usage on writes
-- Comparable compression ratios
```

## Auto-fix
This rule is not auto-fixable. Replace pglz with lz4 in SET COMPRESSION statements. Requires PostgreSQL 14 or later.

## Configuration
```yaml
rules:
  MP077: false           # disable
  MP077:
    severity: critical   # upgrade from warning to critical
```

## Links
- [PostgreSQL ALTER TABLE — SET COMPRESSION](https://www.postgresql.org/docs/current/sql-altertable.html)
- [PostgreSQL TOAST — Compression](https://www.postgresql.org/docs/current/storage-toast.html)
