# MP072: warn-partition-default-scan

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
ALTER TABLE ... ATTACH PARTITION when a DEFAULT partition exists, which triggers a full scan of the default partition under an ACCESS EXCLUSIVE lock.

## Why It's Dangerous
When attaching a new partition, PostgreSQL must verify that no rows in the DEFAULT partition belong to the new partition's range. This requires scanning the entire DEFAULT partition while holding an ACCESS EXCLUSIVE lock on the partitioned table, blocking all reads and writes. On large default partitions, this scan can take minutes or hours. The safe approach is to move qualifying rows out of the DEFAULT partition first, then attach the new partition with a lock_timeout to fail fast if the lock cannot be acquired quickly.

## Bad Example
```sql
ALTER TABLE events ATTACH PARTITION events_2024
  FOR VALUES FROM ('2024-01-01') TO ('2025-01-01');
-- Scans entire DEFAULT partition under ACCESS EXCLUSIVE lock
-- Blocks all reads and writes during scan
-- Can take hours on large default partitions
```

## Good Example
```sql
-- Step 1: Move rows from DEFAULT partition that belong to the new range
BEGIN;
DELETE FROM events_default WHERE event_date >= '2024-01-01' AND event_date < '2025-01-01';
INSERT INTO events_2024 SELECT * FROM events_default WHERE event_date >= '2024-01-01' AND event_date < '2025-01-01';
COMMIT;

-- Step 2: Attach with lock_timeout to fail fast
SET lock_timeout = '3s';
ALTER TABLE events ATTACH PARTITION events_2024
  FOR VALUES FROM ('2024-01-01') TO ('2025-01-01');
```

## Auto-fix
This rule is not auto-fixable. Move qualifying rows out of the DEFAULT partition before attaching the new partition. Use lock_timeout to prevent long-running locks.

## Configuration
```yaml
rules:
  MP072: false           # disable
  MP072:
    severity: critical   # upgrade from warning to critical
```

## Links
- [PostgreSQL ALTER TABLE â€” ATTACH PARTITION](https://www.postgresql.org/docs/current/sql-altertable.html)
- [PostgreSQL Table Partitioning](https://www.postgresql.org/docs/current/ddl-partitioning.html)
