# MP073: ban-superuser-role

**Severity**: Critical | **Auto-fixable**: No | **Tier**: Free

## What It Detects
Migration statements that require superuser privileges, such as ALTER SYSTEM, CREATE ROLE ... WITH SUPERUSER, or other superuser-only operations.

## Why It's Dangerous
Migrations should run with the minimum privileges necessary. ALTER SYSTEM modifies postgresql.conf for the entire cluster, affecting all databases and connections â€” not just the target database. CREATE ROLE WITH SUPERUSER grants unrestricted access to the entire PostgreSQL cluster, bypassing all permission checks. If a migration tool or CI pipeline has superuser access, a compromised migration file could exfiltrate data, drop databases, or modify authentication settings. Use database-level settings and least-privilege roles instead.

## Bad Example
```sql
ALTER SYSTEM SET max_connections = '200';
-- Modifies the entire cluster
-- Requires superuser privileges
-- Takes effect on next server restart

CREATE ROLE admin WITH SUPERUSER;
-- Grants unrestricted access to entire cluster
-- Bypasses all permission checks
```

## Good Example
```sql
-- Use database-level settings instead of cluster-wide
ALTER DATABASE mydb SET max_connections = '200';

-- Create roles with minimal privileges
CREATE ROLE reader WITH LOGIN;
GRANT SELECT ON ALL TABLES IN SCHEMA public TO reader;
```

## Auto-fix
This rule is not auto-fixable. Replace ALTER SYSTEM with database-level ALTER DATABASE settings. Use least-privilege roles instead of SUPERUSER.

## Configuration
```yaml
rules:
  MP073: false           # disable
  MP073:
    severity: warning    # downgrade from critical to warning
```

## Links
- [PostgreSQL ALTER SYSTEM](https://www.postgresql.org/docs/current/sql-altersystem.html)
- [PostgreSQL CREATE ROLE](https://www.postgresql.org/docs/current/sql-createrole.html)
- [PostgreSQL Database Roles](https://www.postgresql.org/docs/current/user-manag.html)
