# MP074: require-deferrable-fk

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
Foreign key constraints created without the DEFERRABLE clause.

## Why It's Dangerous
Non-deferrable foreign keys are checked immediately on each row during INSERT, UPDATE, and DELETE operations. This makes bulk operations significantly slower because each row triggers a lookup on the referenced table. It also makes certain data loading patterns impossible â€” for example, inserting rows into two tables with mutual foreign keys. DEFERRABLE INITIALLY DEFERRED delays constraint checking until COMMIT, allowing bulk operations to complete without per-row validation overhead and enabling circular reference patterns.

## Bad Example
```sql
ALTER TABLE orders ADD CONSTRAINT fk
  FOREIGN KEY (user_id) REFERENCES users (id);
-- Checked on every single row insert
-- Slows bulk operations significantly
-- Cannot handle circular references
```

## Good Example
```sql
ALTER TABLE orders ADD CONSTRAINT fk
  FOREIGN KEY (user_id) REFERENCES users (id)
  DEFERRABLE INITIALLY DEFERRED;
-- Checked at COMMIT time instead of per-row
-- Bulk operations run much faster
-- Supports circular reference patterns
```

## Auto-fix
This rule is not auto-fixable. Add DEFERRABLE INITIALLY DEFERRED to foreign key constraints that will be used with bulk operations.

## Configuration
```yaml
rules:
  MP074: false           # disable
  MP074:
    severity: critical   # upgrade from warning to critical
```

## Links
- [PostgreSQL Foreign Keys](https://www.postgresql.org/docs/current/ddl-constraints.html#DDL-CONSTRAINTS-FK)
- [PostgreSQL SET CONSTRAINTS](https://www.postgresql.org/docs/current/sql-set-constraints.html)
