# MP076: warn-xid-consuming-retry

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
SAVEPOINT statements in migration files, which create subtransactions that consume transaction IDs (XIDs).

## Why It's Dangerous
Each SAVEPOINT creates a subtransaction that consumes a transaction ID (XID). PostgreSQL uses 32-bit XIDs with a wraparound limit of approximately 2 billion. In high-throughput systems, heavy use of subtransactions accelerates XID consumption and brings the database closer to wraparound â€” a catastrophic event where PostgreSQL must shut down to prevent data corruption. Subtransactions also hold onto their snapshot longer, preventing VACUUM from reclaiming dead tuples. Use separate top-level transactions instead of savepoints to avoid unnecessary XID consumption.

## Bad Example
```sql
SAVEPOINT my_savepoint;
-- Consumes a transaction ID
-- Accelerates XID wraparound risk
-- Prevents VACUUM from reclaiming dead tuples
-- Subtransaction overhead on every SAVEPOINT
```

## Good Example
```sql
-- Use separate transactions instead of subtransactions
COMMIT;
BEGIN;
-- ... next operation ...
COMMIT;

-- Or use advisory locks for coordination
SELECT pg_advisory_lock(12345);
```

## Auto-fix
This rule is not auto-fixable. Restructure the migration to use separate top-level transactions instead of savepoints. Monitor XID consumption with pg_database.datfrozenxid.

## Configuration
```yaml
rules:
  MP076: false           # disable
  MP076:
    severity: critical   # upgrade from warning to critical
```

## Links
- [PostgreSQL SAVEPOINT](https://www.postgresql.org/docs/current/sql-savepoint.html)
- [PostgreSQL Transaction ID Wraparound](https://www.postgresql.org/docs/current/routine-vacuuming.html#VACUUM-FOR-WRAPAROUND)
