# MP071: ban-rename-in-use-column

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
ALTER TABLE ... RENAME COLUMN, which renames an existing column in place.

## Why It's Dangerous
Renaming a column immediately breaks all views, functions, triggers, and application queries that reference the old column name. PostgreSQL does not automatically update dependent objects â€” views will fail with "column does not exist" errors, and functions referencing the old name will raise runtime exceptions. The safe approach is an add-copy-drop pattern: add a new column with the desired name, copy data, update all dependents, then drop the old column in a later migration.

## Bad Example
```sql
ALTER TABLE users RENAME COLUMN name TO full_name;
-- Immediately breaks all views referencing "name"
-- Functions and triggers fail at runtime
-- Application queries return errors
```

## Good Example
```sql
-- Step 1: Add new column
ALTER TABLE users ADD COLUMN full_name TEXT;

-- Step 2: Copy data
UPDATE users SET full_name = name;

-- Step 3: Update views, functions, triggers to use full_name
-- Step 4: Update application code
-- Step 5: Drop old column in a later migration
ALTER TABLE users DROP COLUMN name;
```

## Auto-fix
This rule is not auto-fixable. Use the add-copy-drop pattern instead of direct rename. Update all dependent objects (views, functions, triggers) before dropping the old column.

## Configuration
```yaml
rules:
  MP071: false           # disable
  MP071:
    severity: critical   # upgrade from warning to critical
```

## Links
- [PostgreSQL ALTER TABLE](https://www.postgresql.org/docs/current/sql-altertable.html)
- [PostgreSQL Dependency Tracking](https://www.postgresql.org/docs/current/ddl-depend.html)
