# MP078: warn-extension-version-pin

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
CREATE EXTENSION without a VERSION clause, which installs the default (latest) version of the extension.

## Why It's Dangerous
Without an explicit VERSION clause, CREATE EXTENSION installs whatever version is the default on the current PostgreSQL server. This means the same migration file can produce different results on different environments â€” development might get version 1.3 while production gets version 1.2. Extension version mismatches can cause subtle behavior differences, missing functions, or migration failures. Pinning the version ensures reproducible migrations across all environments and makes extension upgrades an explicit, reviewable change.

## Bad Example
```sql
CREATE EXTENSION IF NOT EXISTS pgcrypto;
-- Installs whatever version is default on the server
-- Different environments may get different versions
-- Non-reproducible migrations
```

## Good Example
```sql
CREATE EXTENSION IF NOT EXISTS pgcrypto VERSION '1.3';
-- Explicit version ensures reproducibility
-- Same version across all environments
-- Upgrades are explicit and reviewable
```

## Auto-fix
This rule is not auto-fixable. Add a VERSION clause to CREATE EXTENSION statements. Check the current version with SELECT * FROM pg_available_extensions WHERE name = 'extension_name'.

## Configuration
```yaml
rules:
  MP078: false           # disable
  MP078:
    severity: critical   # upgrade from warning to critical
```

## Links
- [PostgreSQL CREATE EXTENSION](https://www.postgresql.org/docs/current/sql-createextension.html)
- [PostgreSQL Extension System](https://www.postgresql.org/docs/current/extend-extensions.html)
