# MP079: warn-rls-policy-completeness

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
Row-Level Security (RLS) enabled on a table with policies that do not cover all operations (SELECT, INSERT, UPDATE, DELETE).

## Why It's Dangerous
When RLS is enabled on a table, any operation without a matching policy is silently denied — no error is raised, the rows are simply invisible or the operation silently fails. If you create a SELECT policy but forget INSERT, UPDATE, and DELETE policies, those operations will silently do nothing for non-superuser roles. This leads to data loss scenarios where application code believes it successfully inserted or updated rows when nothing actually happened. Covering all operations or using FOR ALL ensures no silent denials.

## Bad Example
```sql
ALTER TABLE posts ENABLE ROW LEVEL SECURITY;

CREATE POLICY posts_select ON posts FOR SELECT USING (true);
-- INSERT, UPDATE, DELETE silently denied for non-superusers
-- No error raised — operations appear to succeed
-- Data loss: inserts/updates silently discarded
```

## Good Example
```sql
ALTER TABLE posts ENABLE ROW LEVEL SECURITY;

-- Option 1: Create policies for ALL operations
CREATE POLICY posts_select ON posts FOR SELECT USING (true);
CREATE POLICY posts_insert ON posts FOR INSERT WITH CHECK (auth.uid() = user_id);
CREATE POLICY posts_update ON posts FOR UPDATE USING (auth.uid() = user_id);
CREATE POLICY posts_delete ON posts FOR DELETE USING (auth.uid() = user_id);

-- Option 2: Use FOR ALL to cover every operation
CREATE POLICY posts_all ON posts FOR ALL USING (auth.uid() = user_id);
```

## Auto-fix
This rule is not auto-fixable. Create policies for all operations (SELECT, INSERT, UPDATE, DELETE) or use FOR ALL. Review existing policies with: SELECT * FROM pg_policies WHERE tablename = 'table_name'.

## Configuration
```yaml
rules:
  MP079: false           # disable
  MP079:
    severity: critical   # upgrade from warning to critical
```

## Links
- [PostgreSQL Row Security Policies](https://www.postgresql.org/docs/current/ddl-rowsecurity.html)
- [PostgreSQL CREATE POLICY](https://www.postgresql.org/docs/current/sql-createpolicy.html)
