# MP080: ban-data-in-migration

**Severity**: Warning | **Auto-fixable**: No | **Tier**: Free

## What It Detects
Data manipulation statements (INSERT, UPDATE, DELETE) in DDL migration files that also contain schema changes.

## Why It's Dangerous
Mixing schema changes (CREATE TABLE, ALTER TABLE) with data changes (INSERT, UPDATE, DELETE) in the same migration file creates several problems. Schema changes typically acquire strong locks and should complete as quickly as possible â€” data manipulation extends the lock duration. If the data manipulation fails partway through, the schema change may have already been applied, leaving the database in an inconsistent state. Separating schema and data changes into different migration files allows independent rollback, makes migrations easier to review, and keeps lock durations minimal.

## Bad Example
```sql
CREATE TABLE settings (key TEXT, value TEXT);
INSERT INTO settings VALUES ('version', '1');
-- Schema change and data change in same migration
-- If INSERT fails, CREATE TABLE already applied
-- Lock held longer than necessary
-- Harder to review and rollback independently
```

## Good Example
```sql
-- Migration 001: Schema change only
CREATE TABLE settings (key TEXT, value TEXT);

-- Migration 002: Data seeding (separate file)
INSERT INTO settings VALUES ('version', '1');
```

## Auto-fix
This rule is not auto-fixable. Separate schema changes and data manipulation into different migration files. Use your migration framework's ordering to ensure schema migrations run before data migrations.

## Configuration
```yaml
rules:
  MP080: false           # disable
  MP080:
    severity: critical   # upgrade from warning to critical
```

## Links
- [PostgreSQL INSERT](https://www.postgresql.org/docs/current/sql-insert.html)
- [PostgreSQL CREATE TABLE](https://www.postgresql.org/docs/current/sql-createtable.html)
